Add initial content for explanantion/development process (#54)
canonical/ubuntu-packaging-guide: docs/explanation/development-process.rst

Ubuntu development process
==========================

Each release cycle follows the same general pattern, with the following
major phases. Ubuntu contributors are expected to follow this process closely
to ensure that their work is aligned with that of others. Because of the

> time-based release cycle, Ubuntu contributors must be well-coordinated to

Reviewer: ```suggestion
time-based release cycle, Ubuntu contributors must coordinate well to
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

produce an on-time release.

See also the article, :doc:`/explanation/releases`, for more details about the
release cadence.

> Beginning a new Release

Reviewer: ```suggestion
Beginning a new release
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

-----------------------

The Ubuntu infrastructure is prepared for a new development branch at the
beginning of each cycle. The package build system is set up, the toolchain

> is organized, :term:`seeds` are branched, and many other pieces are made ready

Reviewer: ```suggestion
is organised, :term:`seeds` are branched, and many other pieces are made ready
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

before development can properly begin. Once these preparations are made, the
new branch is officially announced on the
`ubuntu-devel-announce mailing list <https://lists.ubuntu.com/mailman/listinfo/ubuntu-devel-announce>`_ 
and opened for uploads to the :doc:`/explanation/archive`.

.. note::
    See the `Ubuntu 24.04 LTS (Noble Numbat) archive opening announcement email <https://lists.ubuntu.com/archives/ubuntu-devel-announce/2023-October/001341.html>`_

>     as an example announcement for an archive opening.

Reviewer: ```suggestion
    as an example.
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

Planning
--------

Ubuntu contributors discuss the targeted features for each release cycle via
the various channels (e.g., IRC, Matrix, Discourse, Launchpad). Some of these
come from strategic priorities for the distribution as a whole, and some are
proposed by individual developers.

The broader open-source community gets together at the :term:`Ubuntu Summit` 
(similar but different to the past 
:term:`Ubuntu Developer Summits <Ubuntu Developer Summit>`) to share
experiences and ideas and to inspire future exciting projects covering
development as well as design, writing, and community leadership with a wide
range of technical skill levels.

The :term:`Canonical` teams meet after every release at the Roadmap Planning 
Sprint to plan the roadmap for the next release and discuss how to implement
the roadmap at the Engineering Sprint.

> Merging with Upstream & Feature Development
> -------------------------------------------

Reviewer: ```suggestion
Merging with upstream and feature development
---------------------------------------------
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

> The first phase of the release cycle is characterized by bringing new releases

Reviewer: ```suggestion
The first phase of the release cycle is characterised by bringing new releases
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

of :term:`upstream` components into Ubuntu, either directly or via 
:doc:`Merges and Syncs from Debian </explanation/debian-merges-and-syncs>`. 
Development of planned projects for the release often begins while merging is
still underway, and their development is accelerated once the package archive
is reasonably consistent and usable.

The automatic import of new package versions from Debian ends at the 
:ref:`DebianImportFreeze`.

> Stabilization & Deadlines (Freezes)
> -----------------------------------

Reviewer: ```suggestion
Stabilisation and deadlines (freezes)
-------------------------------------
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

> During development, greater caution is gradually exercised in making changes

Reviewer: ```suggestion
During development, caution is increasingly exercised in making changes
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

to Ubuntu to ensure a stable state is reached in time for the final release

> date. The typical order and length of the various freezes can be seen on the

Reviewer: It's not quite clear what the definition of a freeze is here. I think we could be a bit more explicit about what we mean by "freeze", to avoid using the term without making it really clear what the term means.

current Release Schedule, which is usually posted as a Discourse article under
the `"Release" topic <https://discourse.ubuntu.com/c/release/>`_. 

.. note::
    In the past, the Release Schedule was published in the Ubuntu Wiki.
    See for example the `release schedule of Ubuntu 20.04 LTS (FocalFossa) <https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule>`_.

During freezes, exceptions must be requested to approve changes.
See :doc:`how to request a freeze exception </how-to/request-freeze-exception>`.

.. _TestingWeeks:

> Testing Weeks

Reviewer: ```suggestion
Testing weeks
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

~~~~~~~~~~~~~

> During a release's development phase, testing weeks are organized to focus the

Reviewer: ```suggestion
During a release's development phase, testing weeks are organised to focus the
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

Ubuntu community's efforts on testing Ubuntu's latest daily
:term:`ISO images <Image>` and its :term:`flavours <Ubuntu flavours>`. These
weeks are crucial for discovering bugs and getting early feedback about new
features.

.. note::
    The testing weeks replaced the older practice of alpha and beta milestones.

>     For example, Ubuntu 14.04 LTS (Trusty Tahr) had Alpha 1, Alpha 2, Beta 1,

Reviewer: ```suggestion
    For example, Ubuntu 14.04 LTS (Trusty Tahir) had Alpha 1, Alpha 2, Beta 1,
```

Author: "Tahr" is correct according to:
- https://www.releases.ubuntu.com/trusty/
- https://wiki.ubuntu.com/Releases
- https://wiki.ubuntu.com/TrustyTahr/ReleaseNotes

>     and Beta 2 Milestones.

Reviewer: ```suggestion
    and Beta 2 milestones.
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

    See `the email <https://lists.ubuntu.com/archives/ubuntu-release/2018-April/004434.html>`_
    that announced the process change.

.. _DebianImportFreeze:

Debian Import Freeze
~~~~~~~~~~~~~~~~~~~~

The automatic import of new packages and versions of existing packages from
Debian gets disabled. The import of a new package or version of an existing
package from Debian has to be requested. 

.. note::

    The general development activity is still unrestricted until the
    Feature Freeze; however, the Feature Freeze is often scheduled for the same
    day.

.. _FeatureFreeze:

Feature Freeze (FF)
~~~~~~~~~~~~~~~~~~~

At this point, Ubuntu developers should stop introducing new features,

> packages, and :term:`API`/:term:`ABI` changes and concentrate on fixing bugs

Reviewer: ```suggestion
packages, and :term:`API`/:term:`ABI` changes, and instead concentrate on fixing bugs
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

in the current release in development.

.. _User Interface Freeze:

User Interface Freeze (UIF)
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The user interface should be finalized to allow documentation writers and

Reviewer: ```suggestion
The user interface should be finalised to allow documentation writers and
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

bugfix changes are expected.

.. note::
    The Kernel Feature Freeze occurs after the :ref:`FeatureFreeze` because
    the Linux Kernel is typically released upstream after the Feature Freeze.
    Additionally, the Kernel Feature Freeze is deliberately scheduled so that
    the Beta images have a fully featured kernel suitable for testing. 

.. _HardwareEnablementFreeze:

Hardware Enablement Freeze
~~~~~~~~~~~~~~~~~~~~~~~~~~

All new hardware enablement tasks for devices targeting the given release
should be finished, and all the respective packages should be in the Ubuntu
package archive. The release team will no longer accept changes in the Ubuntu
package archive related to supporting new image types or platforms.
This freeze ensures that any new platforms are already available for testing
of the beta images and in the weeks leading to the :ref:`FinalFreeze`.

.. note::
    The Hardware Enablement Freeze is usually scheduled for the same day as
    the Beta Freeze.

.. _BetaFreeze:

Beta Freeze
~~~~~~~~~~~

For the beta release's preparation, all uploads are queued and subject to
manual approval by the release team. Changes to packages that affect beta

> release images (flavours included) require release team approval before

Reviewer: ```suggestion
release images (flavours included) require the release team's approval before
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

uploading. Uploads for packages that do not affect images will generally be
accepted as time permits.

.. tip::
    You can use the :manpage:`seeded-in-ubuntu(1)` tool, provided by the
    ``ubuntu-dev-tools`` package, to list all the current daily images
    containing a specified package or to determine whether the specified
    package is part of the supported seed. 
    

>     If the outputted list is empty, uploading it during a freeze should be

Reviewer: ```suggestion
    If the list output is empty, uploading it during a freeze should be
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

    safe.

The freeze allows Archive Admins to fix package inconsistencies or critical
bugs quickly and in an isolated manner. Once the beta release is shipped, the 
Beta Freeze restrictions no longer apply.

.. _KernelFreeze:

Kernel Freeze
~~~~~~~~~~~~~

The Kernel Freeze is a deadline for kernel updates since they require several
lockstep actions that must be folded into the image-building process.

Exceptional circumstances may justify exemptions to the freeze at the
discretion of the release managers.

.. _NonLanguagePackTranslationDeadline:

> Non Language Pack Translation Deadline

Reviewer: ```suggestion
Non-language-pack translation deadline
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

release pocket until the Final Release will become candidates for
:ref:`StableReleaseUpdates_Summary`. Therefore, uploads to the proposed pocket
during Final Freeze should meet the requirements of Stable Release Updates if
the upload is not accepted into the release pocket. In particular, the upload
must reference at least one bug, which will be used to track the stable update. 

If you are sure that your upload will be accepted during Final Freeze, you can
upload directly to the release pocket, but be aware that you have to re-upload
after Final Release if the upload gets rejected.

.. _ReleaseCandidate:

Release Candidate
~~~~~~~~~~~~~~~~~

The images produced during the week before the :ref:`FinalRelease` are
considered "release candidates". In an ideal world, the first release candidate
would end up being the Final Release; however, we don't live in a perfect
world, and this week is used to get rid of the last release-critical bugs and
do as much testing as possible. Until the Final Release, changes are only
permitted at the release team's discretion and will only be allowed for
high-priority bugs that might justify delaying the release.

.. _LanguagePackTranslationDeadline:

> Language Pack Translation Deadline

Reviewer: ```suggestion
Language pack translation deadline
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Translations done up until this date will be included in the final release's
language packs. 

> Finalization

Reviewer: ```suggestion
Finalisation
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582

------------

As the final release approaches, the focus narrows to fixing "showstopper"
bugs and thoroughly validating the installation images. Every image is tested
to ensure that the installation methods work as advertised. Low-impact bugs

> and other issues are deprioritized to focus developers on this effort.

Reviewer: ```suggestion
and other issues are deprioritised to focus developers on this effort.
```

Author: Incorporated with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/701d0ece4c74f1ccf5b0d8e157ce4d7cf3e1c582
